src/huffman.rs

```diff
@@ -17,49 +17,45 @@ enum HuffmanTreeNode {
     Empty,
 }
 
-/// Huffman tree
-#[derive(Clone, Debug, Default)]
-pub(crate) struct HuffmanTree {
-    tree: Vec<HuffmanTreeNode>,
-    max_nodes: usize,
-    num_nodes: usize,
+#[derive(Clone, Debug)]
+enum HuffmanTreeInner {
+    Single(u16),
+    Tree {
+        tree: Vec<HuffmanTreeNode>,
+        table: Vec<u32>,
+        table_mask: u16,
+    },
 }
 
-impl HuffmanTree {
-    fn is_full(&self) -> bool {
-        self.num_nodes == self.max_nodes
-    }
-
-    /// Turns a node from empty into a branch and assigns its children
-    fn assign_children(&mut self, node_index: usize) -> usize {
-        let offset_index = self.num_nodes - node_index;
-        self.tree[node_index] = HuffmanTreeNode::Branch(offset_index);
-        self.num_nodes += 2;
+/// Huffman tree
+#[derive(Clone, Debug)]
+pub(crate) struct HuffmanTree(HuffmanTreeInner);
 
-        offset_index
+impl Default for HuffmanTree {
+    fn default() -> Self {
+        Self(HuffmanTreeInner::Single(0))
     }
+}
 
-    /// Init a huffman tree
-    fn init(num_leaves: usize) -> Result<HuffmanTree, DecodingError> {
-        if num_leaves == 0 {
-            return Err(DecodingError::HuffmanError);
-        }
+impl HuffmanTree {
+    /// Builds a tree implicitly, just from code lengths
+    pub(crate) fn build_implicit(code_lengths: Vec<u16>) -> Result<HuffmanTree, DecodingError> {
+        let mut num_symbols = 0;
+        let mut root_symbol = 0;
 
-        let max_nodes = 2 * num_leaves - 1;
-        let tree = vec![HuffmanTreeNode::Empty; max_nodes];
-        let num_nodes = 1;
+        for (symbol, length) in code_lengths.iter().enumerate() {
+            if *length > 0 {
+                num_symbols += 1;
+                root_symbol = symbol.try_into().unwrap();
+            }
+        }
```

Comment:
I wonder if it's beneficial/feasible to combine some of the later `code_lengths.iter()` logic in this function into one loop.

The Vec would be traversed fewer times and do fewer tests of `length != 0`/`length > 0`.

Reply:
Yeah, there's a lot of followup optimizations to do. For instance, a better `read_symbol_slowpath` implementation could probably cut ~5% of total decoding time